### ISSUE TITLE
CollisionPolygon2D keeps reporting collisions when disabled if polygon property is changed

### ISSUE THREAD

**Issue by DualMatrix** [godotengine/godot#17812]
**Godot version:**
3.0.2

**Issue description:**

An Area2D with a CollisionPolygon2D keeps reporting collisions even when the CollisionPolygon2D is disabled. This only happens after the _polygon_ property is changed from code.

**Steps to reproduce:**

Make an Area2D, attach a CollisionPolygon2D to it. Disable the CollisionPolygon2D. Connect a print statement to the body_enter signal of the Area2D. Make sure the _polygon_ property of the CollisionPolygon2D is edited at least once from code, you can simply do $CollisionPolygon2D.polygon = $CollisionPolygon2D.polygon in __ready_ to achieve this.

Create a controllable KinematicBody2D and run the scene.

Move the KinematicBody2D inside the Area and the print statement will be executed, even though collisions should be disabled.

**Minimal reproduction project:**

[debug_collisions_disabled.zip](https://github.com/godotengine/godot/files/1854060/debug_collisions_disabled.zip)

**Comment by ShyRed** [comment]
Might be solved once #17151 is merged.

**Comment by qarmin** [comment]
Bug is still preset on master(a16d9c6).
This is really strange, Godot prints that collision is disabled but still detects it

![zrzut ekranu z 2018-11-10 05-33-00](https://user-images.githubusercontent.com/41945903/48297519-e4dc0d80-e4aa-11e8-9299-d92a66db3ea1.png)

**Comment by Calamander** [comment]
Same thing still happens in **3.1 stable** with **Area2D**/**KinematicBody2D**/**RigidBody2D** owning some **CollisionShape2D** or **CollisionPolygon2D** with **disabled** property set to **true**. After changing shape/polygon properties of said CollisionObject2D (area,kinematic,rigid) from code or animation. This can only be prevented if disabled property toggled false/true right after changing shape/polygon.

**Comment by golddotasksquestions** [comment]
I'm not sure if this is a separate issue or the one at hand here. However I have an animated CollisionPoligon2D that would not register collisions when `.disabled` is animated on and off via AnimationPlayer. (Process mode Physics, on Idle it's even less predictable).
Minimal Project: https://github.com/golddotasksquestions/Collision_Detection-collpoly_disabled

**Comment by antitector** [comment]
Same issue on 3.1.1
It works and doesn't randomly. 
Tried many things, in different scenarios. 
![image](https://user-images.githubusercontent.com/51265270/66785502-5d2a2a80-eed5-11e9-9528-af87ee706f6d.png)

**Comment by golddotasksquestions** [comment]
Have you tried with the new 3.2 alpha2? @polytect

**Comment by polytect** [comment]
> Have you tried with the new 3.2 alpha2? @polytect

Yes i did. Just now. And the problem still the same in '3.2 alpha2' as in the picture below.
What's our next step? 

![image](https://user-images.githubusercontent.com/51265270/66961568-e0c75100-f066-11e9-883b-d3b1cd22ac3d.png)

**Comment by qarmin** [comment]
Can you attach minimal project?
I can't reproduce this on Godot 3.2 master.

**Comment by polytect** [comment]
Trying to catch it. As I tried to reproduce, it didn't worked. Now I'm trying to strip down my project to reproduce it in minimal environment.

**Comment by polytect** [comment]
[Polytects_Collision_bedbugs.zip](https://github.com/godotengine/godot/files/3741556/Polytects_Collision_bedbugs.zip)
![bug](https://user-images.githubusercontent.com/51265270/67053201-84ccfd00-f138-11e9-9d5e-27db7c6f2676.png)

Tested on:
Godot 3.1.1_stable (bug)
Godot 3.2_alpha1 (bug)
Godot 3.2_alpha1 (bug)
System:
GNU/LINUX Debian 10

This one is working bug (roach), was hard to catch.  It is really an unusual situation.
Signal of body entered > function of Body> trough function of node> trough signal back to body. Which works, but fails to execute polygon.

Click play and watch. RED means collision is disabled.
"ui_select" to disable/enable collision directly (works).
Problem might be more simple, but this is exactly my scenario.
And funny thing is that height impacts the collision on my original project, but this is another thing, i will recreate scenario if needed as well. Lets hope it is not a roach nest.

What do you think?

**Comment by polytect** [comment]
`ERROR: body_set_shape_disabled: Can't change this state while flushing queries. Use call_deferred() or set_deferred() to change monitoring state instead.
`
When i used call_deferred() it worked for this.
I is it a bug then? Or something in between. 
Every day something new to learn..
